lmnet/lmnet/datasets/bdd100k.py

```diff
@@ -74,41 +96,71 @@ def __init__(self,
 
         subset_dir = "train" if subset == "train" else "val"
-        self.img_dir = join(self.data_dir, "images", "100k", subset_dir)
-        self.anno_dir = join(self.data_dir, "labels", "100k", subset_dir)
+        self.img_dir = os.path.join(self.data_dir, "images", "100k", subset_dir)
+        self.anno_dir = os.path.join(self.data_dir, "labels", "bdd100k_labels_images_" + subset_dir + ".json")
         self.paths = []
         self.bboxs = []
+        self.save_file = os.path.join("dataset", "bdd100k_%s.pickle" % self.subset)
+        if not os.path.exists(os.path.dirname(self.save_file)):
+            os.makedirs(os.path.dirname(self.save_file))
 
-        self._init_files_and_annotations()
+        if os.path.exists(self.save_file):
+            with open(self.save_file, "rb") as f:
+                loaded_data = pickle.load(f)
+                if len(loaded_data[0]) > 0:
+                    self.paths, self.bboxs = loaded_data
+                else:
+                    os.remove(self.save_file)
+                    self._init_files_and_annotations()
+        else:
+            self._init_files_and_annotations()
 
     def _init_files_and_annotations(self):
-        img_paths = glob.glob(join(self.img_dir, "*.jpg"))
-        anno_paths = [join(self.anno_dir, splitext(basename(f))[0] + ".json")
-                      for f in img_paths]
-
-        self.paths = img_paths
-        for f in anno_paths:
-            with open(f) as fp:
-                raw = json.load(fp)
-            objs = raw["frames"][0]["objects"]
-            bbox = []
-            for obj in objs:
-                cat = obj["category"]
-                cat = cat.replace(" ", "_")
-                if cat in self.classes:
-                    cls_idx = self.classes.index(cat)
-                    x1 = int(round(obj["box2d"]["x1"]))
-                    x2 = int(round(obj["box2d"]["x2"]))
-                    y1 = int(round(obj["box2d"]["y1"]))
-                    y2 = int(round(obj["box2d"]["y2"]))
-                    x = x1
-                    y = y1
-                    w = x2 - x1
-                    h = y2 - y1
-
-                    bbox.append([x, y, w, h, cls_idx])
-            bbox = np.array(bbox, dtype=np.int32)
-            self.bboxs.append(bbox)
+        img_paths = OrderedDict([(os.path.basename(path), path) for path in glob.glob(os.path.join(self.img_dir, "*.jpg"))])
+        anno_data = json.load(open(self.anno_dir))
+
+        bbox = [[] for _ in range(len(img_paths))]
+        img_index = list(img_paths.keys())
+
+        total_count = len(img_paths)
+        print("\rGathering annotation data ... ", end="")
+        counts = 0
+        for i, item in enumerate(anno_data):
+            counts += 1
+            for label in item['labels']:
+                class_name = label['category'].replace(' ', '_')
+                # Skip if Label not in images
+                if not item['name'] in img_index:
+                    continue
```

Comment:
Since img_index is a (large) list, item['name'] is compared to elements of img_index one by one (about 70,000 times compared for each image and labels). It seems better to use 'set'.

Reply:
as `img_index` is not need to preserve order `set` can be used for performance gain. I'll update.

Reply:
OK! Thank you!